Implement swift-inspect dump-concurrency for Windows

jakepetroules opened the pull request:
The main pieces remaining are to provide an API in swift_Concurrency.dll to provide the address of the Swift task pointer thread_local variable, or its fully computed offset, since doing so requires access to the address. In this prototype I hardcode '8' for now.

It also needs general cleanup. 32-bit, or a 64-bit build of the tool reading a 32-bit process, will almost certainly not work.

jakepetroules committed:
Implement swift-inspect dump-concurrency for Windows

The main pieces remaining are to provide an API in swift_Concurrency.dll
to provide the address of the Swift task pointer thread_local variable,
or its fully computed offset, since doing so requires access to the
address. In this prototype I hardcode '8' for now.

compnerd reviewed (commented):
Some early comments on this; I realize that this is still not ready yet.

jakepetroules commented:
@swift-ci test

jakepetroules commented on tools/swift-inspect/Sources/swift-inspect/main.swift:
```diff
@@ -140,6 +140,7 @@ internal struct SwiftInspect: ParsableCommand {
     DumpGenericMetadata.self,
     DumpCacheNodes.self,
     DumpArrays.self,
+    DumpConcurrency.self,
```
Done

jakepetroules commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -397,6 +394,140 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+
+    defer {
+      _ = FreeLibrary(ntdll)
```
Done

jakepetroules commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -397,6 +394,140 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+
+  func read(address: swift_addr_t, size: Int) -> UnsafeRawPointer? {
+    guard let buffer = malloc(Int(size)) else { return nil }
+      if !ReadProcessMemory(
```
Done

jakepetroules commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -397,6 +394,140 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+
+    return tasks
+  }
```
Maybe for a follow-on change?

jakepetroules commented:
@swift-ci test

compnerd reviewed (approved):
Some minor comments, but I think that we should be able to get this change in and iterate as well.

jakepetroules committed:
Add additional context to swift-inspect dump-concurrency output

This tool will now print the offset and module name in addition to the symbol. For example, on Windows:

* 1 child task

`--Task 692 - flags=childTask|future|asyncLetTask enqueuePriority=0x15 maxPriority=0x0 address=0x294f469d800
     parent: 0x294f6a8cb80
     async backtrace: $ss15SuspendingClockV7InstantV1loiySbAD_ADtFZ + 0x20 in swift_Concurrency
                      libclang_cas_replay_compilation + 0x1225b00 in swift-build
                      libclang_cas_replay_compilation + 0x12265b0 in swift-build
                      libclang_cas_replay_compilation + 0x1226800 in swift-build
                      $sScs9unfoldingScsyxs5Error_pGxSgyYaKc_tcsAB_pRs_rlufC + 0x3f0 in swift_Concurrency
                      $ss15SuspendingClockV7InstantV9hashValueSivg + 0x110 in swift_Concurrency
                      $ss27withTaskCancellationHandler9operation8onCancel9isolationxxyYaKXE_yyYbXEScA_pSgYitYaKlF + 0xe0 in swift_Concurrency
                      $sScs9unfoldingScsyxs5Error_pGxSgyYaKc_tcsAB_pRs_rlufC + 0x1f0 in swift_Concurrency
                      $ss15SuspendingClockV7InstantV9hashValueSivg + 0xc0 in swift_Concurrency
                      $sScs8IteratorV4nextxSgyYaKF + 0xc0 in swift_Concurrency
                      $sScg8IteratorV8finishedSbvs + 0x100 in swift_Concurrency
                      $sScI4next7ElementQzSgyYaKFTj + 0x70 in swift_Concurrency
                      libclang_cas_replay_compilation + 0x1299940 in swift-build
                      libclang_cas_replay_compilation + 0x130b3a0 in swift-build
                      libclang_cas_replay_compilation + 0x130b610 in swift-build
                      libclang_cas_replay_compilation + 0x130b750 in swift-build
                      libclang_cas_replay_compilation + 0x130f7a0 in swift-build
                      libclang_cas_replay_compilation + 0x130f9e0 in swift-build
                      swift_task_startOnMainActor + 0x170 in swift_Concurrency
     resume function: $ss15SuspendingClockV7InstantV1loiySbAD_ADtFZ + 0x20 in swift_Concurrency
     task allocator: 3272 bytes in 8 chunks

It also fixes an issue where the name of the module in the which the
symbol was located couldn't be retrieved.

al45tair commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -401,6 +405,141 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+
+      precondition(nt.Signature == IMAGE_NT_SIGNATURE)
+
+      // IMAGE_DIRECTORY_ENTRY_TLS == 9
```
We should check `NumberOfRvaAndSizes` before accessing `DataDirectory.9`.

ktoso committed:
Merge branch 'main' into dump-concurrency-windows

ktoso commented:
@swift-ci please smoke test

ktoso commented:
If tests pass, we're happy enough to take this change -- I discussed with @al45tair offline and I have a follow up about the outstanding comment  in a follow up PR https://github.com/swiftlang/swift/pull/89789

finagolfin commented on tools/swift-inspect/Sources/swift-inspect/Operations/DumpConcurrency.swift:
```diff
@@ -10,13 +10,16 @@
 //===----------------------------------------------------------------------===//
 
-#if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
+#if !os(Linux)
```
This change [broke building this tool for Android on the trunk Windows toolchain CI](https://ci-external.swift.org/job/swift-main-windows-toolchain/2157/console), `C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\tools\swift-inspect\Sources\swift-inspect\Operations\DumpConcurrency.swift:261:22: error: cannot find 'strstr' in scope... if let found = strstr($0, upTo) {`. I notice @ktoso is [disabling this file again for Android in a WIP pull](https://github.com/swiftlang/swift/pull/89789/changes#diff-1f53759d7153e7fbceefce01257d5d098cceadc61c252438ccdf7fcd8f206d06R13): should we wait for that, or just add an import here to get Android building again?

finagolfin replied:
Submitted the minimal version needed to fix the build in #89906, along with the linked linux change from #89789.

jakepetroules commented:
@ktoso Thanks so much for taking this over the finish line!

jakepetroules commented:
> The main pieces remaining are to provide an API in swift_Concurrency.dll to provide the address of the Swift task pointer thread_local variable, or its fully computed offset, since doing so requires access to the address. In this prototype I hardcode '8' for now.

Just want to call out that this is still outstanding - if we ever add more TLS vars in swift_Concurrency.dll this could break mysteriously.
